In [1]:
import datetime
import math

import numpy as np
import pandas_datareader.data as web
import yfinance as yfin

yfin.pdr_override()

In [2]:
# Define all initial variables
# start = datetime.date.today()-datetime.timedelta(365*10)
# end = datetime.date.today()
end = datetime.date(2022, 12, 31)
start = end - datetime.timedelta(365 * 10)

prices = web.DataReader(["GE"], start, end)["Adj Close"]
initialPrice = prices[0]
finalPrice = prices[-1]
cashReturn = (finalPrice - initialPrice) * 100

print(
    "With an initial investment of $%.2f, \
    \nthe cash return of this investment would be \
    \n($%.3f - $%.3f) * 100 = $%.3f"
    % (initialPrice * 100, finalPrice, initialPrice, cashReturn)
)

[*********************100%***********************]  1 of 1 completed
With an initial investment of $10388.75,     
the cash return of this investment would be     
($65.349 - $103.887) * 100 = $-3853.821


In [3]:
# Define all initial variables
# start = datetime.date.today()-datetime.timedelta(365*5)
# end = datetime.date.today()
start = datetime.date(2016, 11, 29)
end = datetime.date(2021, 11, 28)

prices = web.DataReader(["META", "CMG"], start, end)["Adj Close"]
initialMETA = prices.META[0]
initialCMG = prices.CMG[0]
finalMETA = prices.META[-1]
finalCMG = prices.CMG[-1]
METAWeight = initialMETA / (initialMETA + initialCMG)
CMGWeight = initialCMG / (initialMETA + initialCMG)

[*********************100%***********************]  2 of 2 completed


In [4]:
print(
    "We have an initial investment in META of $%.2f and in CMG $%.2f"
    % (initialMETA * 100, initialCMG * 100)
)

print(
    "This would make the weights %.2f and %.2f for META and CMG respectively"
    % (METAWeight, CMGWeight)
)

We have an initial investment in META of $12087.00 and in CMG $39511.00
This would make the weights 0.23 and 0.77 for META and CMG respectively


In [6]:
returnMETA = 100 * (finalMETA - initialMETA) / initialMETA
returnCMG = 100 * (finalCMG - initialCMG) / initialCMG

print(
    "This return over this period for Facebook is %.2f%% and %.2f%% for Chipotle"
    % (returnMETA, returnCMG)
)

print(
    "Multiplying these individual returns by their weights gives %.2f (META) and %.2f (CMG)"
    % (returnMETA * METAWeight, returnCMG * CMGWeight)
)

This return over this period for Facebook is 175.60% and 326.94% for Chipotle
Multiplying these individual returns by their weights gives 41.14 (META) and 250.35 (CMG)


In [7]:
returnMETA * METAWeight + returnCMG * CMGWeight

291.4880512024385

## 2. Calculating Portfolio Variance

In [8]:
weights = np.array([0.23, 0.77])
returns = prices.pct_change()
covariance = 252 * returns.cov()

In [9]:
variance = np.dot(weights.T, np.dot(covariance, weights))

# Print the result
print(str(np.round(variance, 6) * 100) + "%")

8.8018%


We determine that the annual variance of our portfolio is 8.80%

In [10]:
returns.var() * 252

CMG     0.136873
META    0.108623
dtype: float64

We can see in above that both stocks' variance is higher than that of our portfolio, and even though CMG has a 77% weighting and a 13.7% variance, our portfolio's variance is more than 5 percentage points lower. This is because, if assets are not perfectly correlated, there will be some diversification benefit to risk by investing in multiple securities.

### Calculating Standard Deviation of the Portfolio

In [11]:
np.round(math.sqrt(variance) * 100, 2)

29.67

## Performing Similar Analysis on Cryptocurrencies

In [12]:
# Define all initial variables
# start = datetime.date.today()-datetime.timedelta(365*5)
# end = datetime.date.today()
start = datetime.date(2017, 11, 29)
end = datetime.date(2022, 11, 28)

prices = web.DataReader(["BTC-USD", "ETH-USD"], start, end)["Adj Close"]
prices = prices.rename(columns={"BTC-USD": "Bitcoin", "ETH-USD": "Ethereum"})

initialBTC = 5 * prices.Bitcoin[0]
initialETH = 100 * prices.Ethereum[0]
finalBTC = 5 * prices.Bitcoin[-1]
finalETH = 100 * prices.Ethereum[-1]
BTCWeight = initialBTC / (initialBTC + initialETH)
ETHWeight = initialETH / (initialBTC + initialETH)

[*********************100%***********************]  2 of 2 completed


In [13]:
print(
    "This would make the weights %.2f and %.2f for Bitcoin and Ethereum respectively"
    % (BTCWeight, ETHWeight)
)

This would make the weights 0.54 and 0.46 for Bitcoin and Ethereum respectively


In [14]:
returnBTC = 100 * (finalBTC - initialBTC) / initialBTC
returnETH = 100 * (finalETH - initialETH) / initialETH

In [15]:
np.round(returnBTC, 5)

66.29866

In [16]:
np.round(returnETH, 5)

179.54681

In [17]:
np.round(returnBTC * BTCWeight + returnETH * ETHWeight, 5)

118.81345

## Calculating Variance for Cryptos

In [18]:
weights = np.array([0.54, 0.46])
returns = prices.pct_change()
covariance = 365 * returns.cov()
variance = np.dot(weights.T, np.dot(covariance, weights))
print(str(np.round(variance, 5) * 100) + "%")

65.999%


In [19]:
np.round(math.sqrt(variance) * 100, 2)

81.24